### ISSUE TITLE
Attachment - User-defined path and filename

### ISSUE THREAD

**Issue by uwunderli** [moneymanagerex/moneymanagerex#6574]
Hello everyone
I would like to be able to personalize the path and file names of the attachments with some variables. Variables would then be e.g:
[year] = year of booking
[Month] = month of booking
[Day] = Day of the booking
[Amount] = Amount
[Payee] = Payee
[Category] = Category (subcategory (:) is replaced by a (/))
etc...

A specially defined path could then look like this:
%DATABASE%[Year]/[Year]-[Month]-[Day] [Payee] [Amount]

The data path and the file would then look like this:
%DATABASE%/2024/2024-02-15 PayPal -18.30.pdf

If there are two files with the same name, you would have to insert a numbering...

With this function you can structure the documents well so that they can also be used outside of mmex. For example, I often have to send a receipt to a public authority, and this makes it easy to search for the receipts and send them with a mail program without having to save them to a different path and rename them beforehand...

**Comment by rmelillo76** [comment]
I think this could be a useful feature for others.  Here's an example of how it's done in the program that I use for screenshots (Greenshot):

![image](https://github.com/moneymanagerex/moneymanagerex/assets/95182117/ef60bec2-18e6-491e-9b36-51aff565b8cb)
https://getgreenshot.org/

A couple of things to think about:

1. The Payees and Category strings would need to be cleansed so that they don't contain restricted path characters as you mentioned before creating folders and files. Did you mean that Sub categories would generate sub-folders? There can be an infinite level of categories, E.g. 
%DATABASE%/2024/ParentCategory/ChildCategory/NextChildCategory/2024-02-15 Payee -18.30.pdf
It may be simpler to use dashes, so there is one 1 folder (see my next point):
%DATABASE%/2024/ParentCategory-ChildCategory-NextChildCategory/2024-02-15 Payee -18.30.pdf



2. Changes to the transaction after the attachment has been saved could be a challenge. The current filename uses the transaction type and transaction ID so it's not an issue if the data were to change.

![image](https://github.com/moneymanagerex/moneymanagerex/assets/95182117/8d5eb11c-4a61-4c5a-80b5-a18fa073d03c)

**Comment by gabriele-v** [comment]
Even if I understand your point, I think we shouldn't overcomplicate MMEX both on user side and on coding side.

There are too many possible drawbacks and checks to be done related to file path/naming (invalid characters, path length etch..) especially on a cross-platform.

What could be done is an option named "save shortcuts to files instad of moving them to attachments folder" which, if enabled, it simply saves the full file path in the filename, basically acting like a shortcut.

Doing so, we will completely make MMEX agnostic of file name checks: if a file exists on the pc it imply that all checks have already been done by the underlying OS.

**Comment by uwunderli** [comment]
Yes, I understand that you don't want to make MMEX too complicated. It's always a balancing act between user-friendliness and practical functions. A link to an existing file would certainly also be a solution.

The advantage of a file structure that you can enter in MMEX would be that the path and the file names always follow the same structure. So you would no longer have to worry about organizing and naming the files.

But as I said, a link to an existing file would also be good.

I got the idea from Paperless, which we also work with. Here is a screenshot of what it looks like.

Translated with DeepL.com (free version)
![Bildschirmfoto vom 2024-03-10 19-44-39](https://github.com/moneymanagerex/moneymanagerex/assets/55878516/6a3a4462-df2e-4b5e-a086-332c6da5df1a)

**Comment by luisdev** [comment]
I think this is still worth consideration.
